File: src/main/java/no/nav/fo/veilarbregistrering/db/arbeidssoker/ArbeidssokerperioderMapper.java
package no.nav.fo.veilarbregistrering.db.arbeidssoker;

import no.nav.fo.veilarbregistrering.arbeidssoker.Arbeidssokerperiode;
import no.nav.fo.veilarbregistrering.arbeidssoker.Arbeidssokerperioder;
import no.nav.fo.veilarbregistrering.arbeidssoker.Formidlingsgruppe;
import no.nav.fo.veilarbregistrering.bruker.Periode;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;

import static java.util.stream.Collectors.toList;
import static no.nav.fo.veilarbregistrering.arbeidssoker.Arbeidssokerperiode.EldsteFoerst.eldsteFoerst;
import static no.nav.fo.veilarbregistrering.db.arbeidssoker.Formidlingsgruppeendring.NyesteFoerst.nyesteFoerst;

class ArbeidssokerperioderMapper {

    static Arbeidssokerperioder map(List<Formidlingsgruppeendring> formidlingsgruppeendringer) {
        return new Arbeidssokerperioder(
                Optional.of(
                        formidlingsgruppeendringer.stream()
                                .sorted(nyesteFoerst())
                                .collect(toList()))
                        .map(beholdKunEndringerForAktiveIdenter)
                        .map(beholdKunSisteEndringPerDagIListen)
                        .map(populerTilDato)
                        .get()
                        .stream()
                        .sorted(eldsteFoerst())
                        .collect(toList()));
    }

    private static Function<List<Formidlingsgruppeendring>, List<Formidlingsgruppeendring>> beholdKunEndringerForAktiveIdenter =
            (formidlingsgruppeendringer) -> formidlingsgruppeendringer.stream()
                    .filter(Formidlingsgruppeendring::erAktiv)
                    .collect(toList());

    private static Function<List<Formidlingsgruppeendring>, List<Arbeidssokerperiode>> beholdKunSisteEndringPerDagIListen =
            (formidlingsgruppeendringer) -> {
                List<Arbeidssokerperiode> arbeidssokerperioder = new ArrayList<>(formidlingsgruppeendringer.size());

                LocalDate forrigeEndretDato = null;

                for (Formidlingsgruppeendring formidlingsgruppeendring : formidlingsgruppeendringer) {
                    LocalDate endretDato = formidlingsgruppeendring.getFormidlingsgruppeEndret().toLocalDateTime().toLocalDate();

                    if (forrigeEndretDato != null && endretDato.isEqual(forrigeEndretDato)) {
                        continue;
                    }

                    arbeidssokerperioder.add(new Arbeidssokerperiode(
                            Formidlingsgruppe.of(formidlingsgruppeendring.getFormidlingsgruppe()),
                            Periode.of(
                                    endretDato,
                                    null
                            )
                    ));

                    forrigeEndretDato = endretDato;
                }

                return arbeidssokerperioder;
            };

    private static Function<List<Arbeidssokerperiode>, List<Arbeidssokerperiode>> populerTilDato =
            (arbeidssokerperioder) -> {
                List<Arbeidssokerperiode> nyListe = new ArrayList(arbeidssokerperioder.size());

                LocalDate nyTildato = null;
                for (Arbeidssokerperiode arbeidssokerperiode : arbeidssokerperioder) {
                    nyListe.add(arbeidssokerperiode.tilOgMed(nyTildato));
                    nyTildato = arbeidssokerperiode.getPeriode().getFra().minusDays(1);
                }
                return nyListe;
            };
}

File: src/main/kotlin/no/nav/fo/veilarbregistrering/db/arbeidssoker/ArbeidssokerperioderMapper.kt
package no.nav.fo.veilarbregistrering.db.arbeidssoker

import no.nav.fo.veilarbregistrering.arbeidssoker.Arbeidssokerperiode
import no.nav.fo.veilarbregistrering.arbeidssoker.Arbeidssokerperioder
import no.nav.fo.veilarbregistrering.arbeidssoker.EldsteFoerst
import no.nav.fo.veilarbregistrering.arbeidssoker.Formidlingsgruppe
import no.nav.fo.veilarbregistrering.bruker.Periode
import no.nav.fo.veilarbregistrering.db.arbeidssoker.Formidlingsgruppeendring.NyesteFoerst
import java.time.LocalDate

internal object ArbeidssokerperioderMapper {
    fun map(formidlingsgruppeendringer: List<Formidlingsgruppeendring>): Arbeidssokerperioder {
        return Arbeidssokerperioder(
                formidlingsgruppeendringer
                    .sortedWith(NyesteFoerst.nyesteFoerst())
                    .filter(Formidlingsgruppeendring::erAktiv)
                    .run(::slettTekniskeISERVEndringer)
                    .run(::beholdKunSisteEndringPerDagIListen)
                    .run(::populerTilDatoMedNestePeriodesFraDatoMinusEn)
                    .run(::tilArbeidssokerperioder)
                    .sortedWith(EldsteFoerst())
        )
    }

    private fun tilArbeidssokerperioder(formidlingsgruppeperioder: List<Formidlingsgruppeperiode>): List<Arbeidssokerperiode> {
        return formidlingsgruppeperioder
            .filter { it.formidlingsgruppe.erArbeidssoker() }
            .map { Arbeidssokerperiode(it.periode) }
    }

    private fun slettTekniskeISERVEndringer(formidlingsgruppeendringer: List<Formidlingsgruppeendring>) =
        formidlingsgruppeendringer.groupBy { it.formidlingsgruppeEndret }
            .values.flatMap { samtidigeEndringer -> if (samtidigeEndringer.size > 1) samtidigeEndringer.filter { !it.erISERV() } else samtidigeEndringer }
            .sortedWith(NyesteFoerst.nyesteFoerst())

    private fun beholdKunSisteEndringPerDagIListen(formidlingsgruppeendringer: List<Formidlingsgruppeendring>): List<Formidlingsgruppeperiode> {
        val formidlingsgruppeperioder: MutableList<Formidlingsgruppeperiode> = mutableListOf()

        var forrigeEndretDato = LocalDate.MAX
        for (formidlingsgruppeendring in formidlingsgruppeendringer) {
            val endretDato = formidlingsgruppeendring.formidlingsgruppeEndret.toLocalDateTime().toLocalDate()
            if (endretDato.isEqual(forrigeEndretDato)) {
                continue
            }
            formidlingsgruppeperioder.add(
                Formidlingsgruppeperiode(
                    Formidlingsgruppe(formidlingsgruppeendring.formidlingsgruppe),
                    Periode(
                        endretDato,
                        null
                    )
                )
            )
            forrigeEndretDato = endretDato
        }

        return formidlingsgruppeperioder
    }

    private fun populerTilDatoMedNestePeriodesFraDatoMinusEn(formidlingsgruppeperioder: List<Formidlingsgruppeperiode>): List<Formidlingsgruppeperiode> =
        formidlingsgruppeperioder.mapIndexed { index, formidlingsgruppeperiode ->
            val forrigePeriodesFraDato = if (index > 0) formidlingsgruppeperioder[index - 1].periode.fra else null
            formidlingsgruppeperiode.tilOgMed(forrigePeriodesFraDato?.minusDays(1))
        }
}

internal data class Formidlingsgruppeperiode (val formidlingsgruppe: Formidlingsgruppe, val periode: Periode) {
    fun tilOgMed(tilDato: LocalDate?): Formidlingsgruppeperiode {
        return of(
            formidlingsgruppe,
            periode.tilOgMed(tilDato)
        )
    }

    companion object {
        fun of(formidlingsgruppe: Formidlingsgruppe, periode: Periode): Formidlingsgruppeperiode {
            return Formidlingsgruppeperiode(formidlingsgruppe, periode)
        }
    }
}